PhysicsTools/PatAlgos/src/SoftMuonMvaEstimator.cc

```diff
@@ -77,18 +77,70 @@ void SoftMuonMvaEstimator::computeMva(const pat::Muon& muon)
 	
 	iValidFraction_ =  iTrack->validFraction();
 	innerChi2_ =  iTrack->normalizedChi2();
-	layersWithMeasurement_ = static_cast<float>(iTrack->hitPattern().trackerLayersWithMeasurement());
+	layersWithMeasurement_ = iTrack->hitPattern().trackerLayersWithMeasurement();
 
 	outerChi2_ = oTrack->normalizedChi2();
 
-	Qprod_ =  static_cast<float>(iTrack->charge()) * static_cast<float>(oTrack->charge());
+	qProd_ =  iTrack->charge()*oTrack->charge();
 
-	int vDT, vCSCmax, vRPC;
-	vDT = gTrack->hitPattern().numberOfValidMuonDTHits();
-	vCSCmax = gTrack->hitPattern().numberOfValidMuonCSCHits();
-	if(vCSCmax>6) vCSCmax=6;
-	vRPC = gTrack->hitPattern().numberOfValidMuonRPCHits();
-	vMuonHitComb_ = (vDT/2 + vCSCmax + vRPC);
+	const reco::HitPattern &gMpattern = gTrack->hitPattern();
+
+	std::vector<int> fvDThits = {0,0,0,0};
+	std::vector<int> fvRPChits = {0,0,0,0};
+	std::vector<int> fvCSChits = {0,0,0,0};
+
+	vMuonHitComb_ = 0;
+
+	for (int i=0;i<gMpattern.numberOfAllHits(reco::HitPattern::TRACK_HITS);i++){
+
+	  uint32_t hit = gMpattern.getHitPattern(reco::HitPattern::TRACK_HITS, i);
+	  if ( !gMpattern.validHitFilter(hit) ) continue;
+
+	  if (gMpattern.getMuonStation(hit) == 1){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[0]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[0]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[0]++;
+	  }
+  	
+	  if (gMpattern.getMuonStation(hit) == 2){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[1]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[1]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[1]++;
+	  }
+
+	  if (gMpattern.getMuonStation(hit) == 3){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[2]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[2]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[2]++;
+	  }
+
+	  if (gMpattern.getMuonStation(hit) == 4){
+	    if (gMpattern.muonDTHitFilter(hit)) fvDThits[3]++;
+	    if (gMpattern.muonRPCHitFilter(hit)) fvRPChits[3]++;
+	    if (gMpattern.muonCSCHitFilter(hit)) fvCSChits[3]++;
+	  }
+
+	}
+  
+
+	if (!((fvDThits.size() < 4)
+	   || (fvRPChits.size() < 4)
+	   || (fvCSChits.size() < 4))){
+
+	  for (unsigned int station = 0; station < 4; ++station) {
+
+	    vMuonHitComb_ += (fvDThits[station])/2.;
```

Comment:
I'm curious why a similar check as done for CSC is not needed here. station 4 has some significant overlaps. 
What's the origin of /2. ? Is it to match max of 12 with 6 from CSC? What about station 4? is it OK for it to have lower weight?

Reply:
The overlap in CSC ME1 was found to be the most significant, with all other contributions being relatively negligible. 
The division by 2 is to match DT with CSC hits, as you assumed.
What do you mean with lower weight for station 4? All stations have the same weight.

Reply:
there are only 8 layers in station 4. Also, having hits in the last station is perhaps more important.

Reply:
Station 4 receives an elevated weight in the segmentCompatibility variable. In this variable such an explicit weight was not desired. Station 4 gets its importance in tracks which leave hits in all stations.